### PR TITLE
feat(Link): add a sbl-frontend Link that uses react-router when needed

### PR DESCRIPTION
**Pull request by billhimmelsbach**
This integrates https://github.com/cfpb/design-system-react/pull/294 into the app, adding a little composition to the DSR component to automatically switch from normal anchor links to `react-router` links depending on the `href`. 

I wanted to make the transition as easy as possible: it's the same props as the DSR `Link`. As long as you're importing the Link from the app instead of the `DSR`, you don't have to worry about figuring out which links should have `isRouterLink` or not.

If we like this approach, I could see us raising this up into the DSR as a "SBL-specific component" in a later PR.

**Note: don't merge this in until https://github.com/cfpb/design-system-react/pull/294 is merged and [this line](https://github.com/cfpb/sbl-frontend/compare/198-integrate-react-router-links?expand=1#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R37) that temporarily updates the DSR is reverted**

## Changes

- adds new `Link` and `ListLink` components to the App that switch between `a` and `RouterLink` depending on what `href` is given
- updates all instances of `Link` / `ListLink` across the app to use the new components
- updates usage of plain `<a>` anchors to use `Link`
- dedupes `react-router-dom` in the vite config that works with this change in the DSR to allow for `react-router` enabled `Links` to work within the same instance of `react-router` (what a pain this was)

## How to test this PR

1. Do external and internal links still work?
2. Do internal links (particularly breadcrumbs) seem faster with less page jank?

## Screenshots
https://github.com/cfpb/sbl-frontend/assets/19983248/a239a808-b200-4244-b910-b6f9141ba244

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -34,7 +34,7 @@
     "@tanstack/react-query": "4.29.7",
     "@tanstack/react-table": "^8.9.1",
     "classnames": "^2.3.2",
-    "design-system-react": "https://github.com/cfpb/design-system-react",
+    "design-system-react": "https://github.com/cfpb/design-system-react#117-react-router-links",
```
Comment (billhimmelsbach) [inline package.json:37]: This line should be reverted [once this PR](https://github.com/cfpb/design-system-react/pull/294) is merged.

File: src/components/Link.tsx [hunk 2]
```diff
@@ -0,0 +1,76 @@
+// this component could potentially be lifted with others into the DSR in a later refactor
```
Comment (meissadia) [inline src/components/Link.tsx:1]: 👍🏾

File: src/components/Link.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  if (url === undefined) {
+    return false;
+  }
+  return url.startsWith('http');
+};
+
+const getIsRouterUsageInferred = (
+  href: string | undefined,
+  isRouterLink: boolean | undefined,
+): boolean => isRouterLink === undefined && !getIsLinkExternal(href);
```
Comment (meissadia) [inline src/components/Link.tsx:20]: Love this idea!

**Review by meissadia — APPROVED** [review]
🚀

**Review by meissadia — APPROVED** [review]
LGTM 👍🏾

**Pull request merged**